cortex-ar-rt-macros/src/lib.rs

```diff
@@ -0,0 +1,491 @@
+///     panic!("On no")
+/// }
+/// ```
+///
+/// The symbol `kmain` is what the assembly code in both the cortex-r-rt and
+/// cortex-a-rt start-up code will jump to, and the `extern "C"` makes it sound
+/// to call from assembly.
+#[proc_macro_attribute]
+pub fn entry(args: TokenStream, input: TokenStream) -> TokenStream {
+    let f = parse_macro_input!(input as ItemFn);
+
+    // check the function signature.
+    //
+    // it should be `fn foo() -> !` or `unsafe fn foo() -> !`
+    let valid_signature = f.sig.constness.is_none()
+        && f.vis == Visibility::Inherited
+        && f.sig.abi.is_none()
+        && f.sig.inputs.is_empty()
+        && f.sig.generics.params.is_empty()
+        && f.sig.generics.where_clause.is_none()
+        && f.sig.variadic.is_none()
+        && match f.sig.output {
+            ReturnType::Default => false,
+            ReturnType::Type(_, ref ty) => matches!(**ty, Type::Never(_)),
+        };
+
+    if !valid_signature {
+        return parse::Error::new(
+            f.span(),
+            "`#[entry]` function must have signature `[unsafe] fn() -> !`",
+        )
+        .to_compile_error()
+        .into();
+    }
+
+    if !args.is_empty() {
+        return parse::Error::new(Span::call_site(), "This attribute accepts no arguments")
+            .to_compile_error()
+            .into();
+    }
+
+    let tramp_ident = Ident::new("__cortex_ar_rt_kmain", Span::call_site());
+    let ident = &f.sig.ident;
+
+    if let Err(error) = check_attr_whitelist(&f.attrs, Kind::Entry) {
+        return error;
+    }
+
+    let (ref cfgs, ref attrs) = extract_cfgs(f.attrs.clone());
+
+    quote!(
+        #(#cfgs)*
+        #(#attrs)*
+        #[doc(hidden)]
+        #[export_name = "kmain"]
+        pub unsafe extern "C" fn #tramp_ident() -> ! {
+            #ident()
+        }
+
+        #f
+    )
+    .into()
+}
+
+/// The set of exceptions we can handle.
+#[derive(Debug, PartialEq)]
+enum Exception {
+    Undefined,
+    SupervisorCall,
+    PrefetchAbort,
+    DataAbort,
+    Irq,
+}
+
+impl std::fmt::Display for Exception {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Exception::Undefined => write!(f, "Undefined"),
+            Exception::SupervisorCall => write!(f, "SupervisorCall"),
+            Exception::PrefetchAbort => write!(f, "PrefetchAbort"),
+            Exception::DataAbort => write!(f, "DataAbort"),
+            Exception::Irq => write!(f, "Irq"),
+        }
+    }
+}
+
+/// Creates an `unsafe` exception handler.
+///
+/// It's `unsafe` because you are not supposed to call it - it should only be
+/// called from assembly routines registered in the interrupt vector table.
+///
+/// When placed on a function like:
+///
+/// ```rust ignore
+/// #[exception(Undefined)]
+/// fn foo(addr: usize) -> ! {
+///     panic!("On no")
+/// }
+/// ```
+///
+/// You get something like:
+///
+/// ```rust
+/// #[export_name = "_undefined_handler"]
```

Comment:
Is there a reason not to add `#[doc(hidden)]` here as well?

Reply:
Oh no, I see I have been inconsistent. That is bad. Should I add it here, or remove it ther?

Reply:
OK, so I marked all the trampolines as `doc(hidden)`. Good catch, thanks.

Reply:
OH i marked it in the example? You're right, I only meant in the code itself.